EIPS/eip-2537.md

```diff
@@ -0,0 +1,293 @@
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiplication
+
+G1 multiplication call expects `160` bytes as an input that is interpreted as byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiexponentiation
+
+G1 multiplication call expects `160*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Any of G1 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 addition
+
+G2 addition call expects `512` bytes as an input that is interpreted as byte concatenation of two G2 points (`256` bytes each). Output is an encoding of addition operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiplication
+
+G2 multiplication call expects `288` bytes as an input that is interpreted as byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiexponentiation
+
+G2 multiplication call expects `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Any of G2 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for pairing
+
+Pairing call expects `384*k` bytes as an inputs that is interpreted as byte concatenation of `k` slices. Each slice has the following structure:
+- `128` bytes of G1 point encoding
+- `256` bytes of G2 point encoding
+
+Output is a `32` bytes where last single byte is `0x01` if pairing result is equal to multiplicative identity in a pairing target field and `0x00` otherwise (which is equivalent of Big Endian encoding of Solidity values `uint256(1)` and `uin256(0)` respectively).
```

Comment:
I think this wording suggests the first 31 bytes can be anything.

Reply:
Current Bn254 precompile uses the same output format, so it's literally `uint256(0/1)`

Reply:
I agre with @axic here. You should state explicitly that it's 31 zeroes followed by `00`, or `01`.

Reply:
Oh, looks like I didn't understand @axic comment on this. Equivalency with `uint256(0/1)` is stated (and it means top 31 byte are zeroes), but I can rephrase it.

Reply:
A precompile shouldn't refer to Solidity types or data structures, because Solidity is a moving target.

Reply:
Updated